### ISSUE TITLE
techniques used in SPEN training

### ISSUE THREAD

**Issue by masashi-y** [cgraber/GSPEN#2]
Hello, thank you sharing great work!
I am reading your codes and have questions especially regarding your SPEN implementation.

In this part of your code where you perform entropic mirror descent (https://github.com/cgraber/GSPEN/blob/master/gspen/models/spen_model.py#L182),
why do you subtract max value from the gradients d(Energy)/dy?
Is this some technique used in entropic mirror descent?
I don't think I find similar snippet in other's SPEN implementations.

And your implementation is different from original SPEN by David Belanger in that
- the loss of yours is just the difference between true and predicted energies, no error term such as squared error (so "loss augmented" inference is not done)
- and it does not clamp the loss to >= 0 as usually done in SVM loss
Am I correct?

Thank you in advance!

**Comment by cgraber** [comment]
Hi, thanks for your interest!

While I can't comment on other implementations, this is done here to ensure numerical stability. Doing this subtraction prevents the exponential from getting too large, and it is mathematically equivalent to the original operation (look at the update equation on page 8 of the [original paper](https://web.iem.technion.ac.il/images/user-files/becka/papers/3.pdf), and consider multiplying/dividing by exp(max_j f'_j))

This implementation does have loss-augmented inference - this is what the "lossaug" term in the input to inference represents. We used loss functions here that could also be used with the classical structured models, meaning they had to be decomposable across regions (e.g. hamming loss, which you can compute as a sum over individual variables)

[This line](https://github.com/cgraber/GSPEN/blob/master/gspen/models/spen_model.py#L316) clamps the loss to >= 0.

**Comment by masashi-y** [comment]
Thank you for the reply.

Maybe the technique is specific to entropic mirror descent?
grad - max grad goes in an unexpected direction in usual unconstrained optimizations such as SGD,
but it works just as scaling since it is enclosed within exp in entropic mirror descent..
Is that your idea or a common practice (is there any reference)??

Sorry, in the above question, I was assuming the application to multi label classification (bibtex and bookmarks).
According to your shell scripts, you do not use either of lossaug or clamping, so the experimental settings is a bit different from Belanger's in this point too (in addition to dropout as noted in your paper)?

**Comment by cgraber** [comment]
You can think of it being similar in spirit to the [log-sum-exp trick](https://en.wikipedia.org/wiki/LogSumExp#log-sum-exp_trick_for_log-domain_calculations), in that it's an implementation of a computation that is mathematically equivalent that prevents numerical overflow. You are correct in saying that it wouldn't work in SGD, but SGD doesn't compute updates using the exponential function so it wouldn't be necessary anyway. 

Your observations are correct, we are not using the exact experimental settings of the original paper (our goal was to produce the best set of results we could using their techniques, not to reproduce them). Regarding the clamping: in practice, the clamping is only necessary if your inference procedure does not find predictions with larger scores than the scores assigned to ground-truth predictions - otherwise, there is no difference between including/excluding it. During our experiments, I did not observe any problems with inference that would cause a negative loss by excluding this clamping (though it is certainly safer to include it).

**Comment by masashi-y** [comment]
Thank you! Everything is super clear now!
